pkg/kernelmonitor/logwatchers/journald/log_watcher.go

```diff
@@ -0,0 +1,163 @@
+// +build journald
+
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package journald
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/coreos/go-systemd/sdjournal"
+	"github.com/golang/glog"
+
+	"k8s.io/node-problem-detector/pkg/kernelmonitor/logwatchers/types"
+	kerntypes "k8s.io/node-problem-detector/pkg/kernelmonitor/types"
+	"k8s.io/node-problem-detector/pkg/kernelmonitor/util"
+)
+
+// Compiling go-systemd/sdjournald needs libsystemd-dev or libsystemd-journal-dev,
+// which is not always available on all os distros and versions.
+// So we add the build tag in this file, so that on unsupported os distro, user can
+// disable this build tag.
+
+// journaldWatcher is the log watcher for journald.
+type journaldWatcher struct {
+	journal *sdjournal.Journal
+	cfg     types.WatcherConfig
+	logCh   chan *kerntypes.KernelLog
+	tomb    *util.Tomb
+}
+
+// NewJournaldWatcher is the create function of journald watcher.
+func NewJournaldWatcher(cfg types.WatcherConfig) types.LogWatcher {
+	return &journaldWatcher{
+		cfg:  cfg,
+		tomb: util.NewTomb(),
+		// A capacity 1000 buffer should be enough
+		logCh: make(chan *kerntypes.KernelLog, 1000),
+	}
+}
+
+// Make sure NewJournaldWatcher is types.WatcherCreateFunc .
+var _ types.WatcherCreateFunc = NewJournaldWatcher
+
+// Watch starts the journal watcher.
+func (j *journaldWatcher) Watch() (<-chan *kerntypes.KernelLog, error) {
+	journal, err := getJournal(j.cfg)
+	if err != nil {
+		return nil, err
+	}
+	j.journal = journal
+	glog.Info("Start watching journald")
+	go j.watchLoop()
+	return j.logCh, nil
+}
+
+// Stop stops the journald watcher.
+func (j *journaldWatcher) Stop() {
+	j.tomb.Stop()
+}
+
+// waitLogTimeout is the timeout waiting for new log.
+const waitLogTimeout = 5 * time.Second
+
+// watchLoop is the main watch loop of journald watcher.
+func (j *journaldWatcher) watchLoop() {
+	defer func() {
+		if err := j.journal.Close(); err != nil {
+			glog.Errorf("Failed to close journal client: %v", err)
+		}
+		j.tomb.Done()
+	}()
+	for {
+		select {
+		case <-j.tomb.Stopping():
+			glog.Infof("Stop watching journald")
+			return
+		default:
+		}
+		// Get next log entry.
+		n, err := j.journal.Next()
```

Comment:
The 5 second `waitLogTimeout` here means that it won't be blocked for too long at least, so this one isn't too bad.